File: android_app/app/src/main/java/com/oceansentinels/app/data/remote/api/OceanSentinelsApi.kt
package com.oceansentinels.app.data.remote.api

import com.oceansentinels.app.data.remote.dto.*
import okhttp3.MultipartBody
import retrofit2.Response
import retrofit2.http.*

/**
 * Retrofit API service interface for Ocean Sentinels backend
 */
interface OceanSentinelsApi {

    // ============= Authentication =============
    
    @FormUrlEncoded
    @POST("auth/login")
    suspend fun login(
        @Field("username") username: String,
        @Field("password") password: String
    ): Response<AuthResponseDto>
    
    @POST("auth/register")
    suspend fun register(
        @Body request: RegisterRequestDto
    ): Response<UserDto>
    
    // ============= Users =============
    
    @GET("users/me")
    suspend fun getCurrentUser(): Response<UserDto>
    
    @PUT("users/me")
    suspend fun updateCurrentUser(
        @Body request: UserUpdateRequestDto
    ): Response<UserDto>
    
    @GET("users/")
    suspend fun getAllUsers(): Response<List<UserDto>>
    
    @GET("users/{id}")
    suspend fun getUser(
        @Path("id") id: Int
    ): Response<UserDto>
    
    @DELETE("users/{id}")
    suspend fun deleteUser(
        @Path("id") id: Int
    ): Response<Unit>
    
    @POST("users/create")
    suspend fun adminCreateUser(
        @Body request: RegisterRequestDto
    ): Response<UserDto>
    
    // ============= Incidents =============
    
    @GET("incidents/")
    suspend fun getIncidents(
        @Query("page") page: Int = 1,
        @Query("size") size: Int = 20,
        @Query("status") status: String? = null,
        @Query("hazard_type") hazardType: String? = null,
        @Query("urgency") urgency: String? = null,
        @Query("search") search: String? = null
    ): Response<IncidentListResponseDto>
    
    @GET("incidents/{id}")
    suspend fun getIncident(
        @Path("id") id: Int
    ): Response<IncidentDto>
    
    @POST("incidents/")
    suspend fun createIncident(
        @Body request: CreateIncidentRequestDto
    ): Response<IncidentDto>

    /**
     * Bulk check which mesh message IDs have already been delivered to the server.
     * Called when a device comes online to discover messages already uploaded by
     * other mesh peers. Those can be marked DELIVERED locally and stop relaying.
     */
    @POST("incidents/mesh/check")
    suspend fun checkMeshMessages(
        @Body request: MeshCheckRequestDto
    ): Response<MeshCheckResponseDto>
    
    @PUT("incidents/{id}/verify")
    suspend fun verifyIncident(
        @Path("id") id: Int
    ): Response<MessageResponseDto>
    
    @PUT("incidents/{id}/deploy")
    suspend fun deployResponse(
        @Path("id") id: Int
    ): Response<MessageResponseDto>
    
    @PUT("incidents/{id}/resolve")
    suspend fun resolveIncident(
        @Path("id") id: Int
    ): Response<MessageResponseDto>
    
    @PUT("incidents/{id}/assign")
    suspend fun assignIncident(
        @Path("id") id: Int,
        @Body body: Map<String, Int>
    ): Response<MessageResponseDto>
    
    @GET("incidents/assigned/me")
    suspend fun getMyAssignedIncidents(
        @Query("page") page: Int = 1,
        @Query("size") size: Int = 20
    ): Response<IncidentListResponseDto>
    
    // ============= Analytics =============
    
    @GET("analytics/dashboard")
    suspend fun getDashboardAnalytics(): Response<DashboardAnalyticsDto>
    
    @GET("analytics/incidents/timeline")
    suspend fun getIncidentsTimeline(
        @Query("days") days: Int = 30
    ): Response<IncidentsTimelineDto>
    
    @GET("analytics/incidents/distribution")
    suspend fun getIncidentsDistribution(): Response<IncidentsDistributionDto>
    
    @GET("analytics/geographic")
    suspend fun getGeographicAnalytics(): Response<GeographicAnalyticsDto>
    
    // ============= File Upload =============
    
    @Multipart
    @POST("upload/incident-photo")
    suspend fun uploadIncidentPhoto(
        @Part photo: MultipartBody.Part
    ): Response<UploadResponseDto>
}

/**
 * Response for file upload
 */
data class UploadResponseDto(
    val url: String,
    val filename: String
)

File: IOS/OceanSentinels/Data/Remote/API/OceanSentinelsAPI.swift
import Foundation

// MARK: - Ocean Sentinels API Service

/// Replaces Retrofit OceanSentinelsApi interface with URLSession-based async service.
final class OceanSentinelsAPI {
    
    private let client: NetworkClient
    private let baseURL: URL
    
    init(client: NetworkClient = .shared) {
        self.client = client
        guard let url = URL(string: AppConfig.baseURL) else {
            fatalError("Invalid base URL: \(AppConfig.baseURL)")
        }
        self.baseURL = url
    }
    
    private func url(_ path: String, queryItems: [URLQueryItem]? = nil) -> URL {
        var components = URLComponents(url: baseURL.appendingPathComponent(path), resolvingAgainstBaseURL: false)!
        if let items = queryItems?.filter({ $0.value != nil }) {
            components.queryItems = items.isEmpty ? nil : items
        }
        return components.url!
    }
    
    // MARK: - Authentication
    
    func login(username: String, password: String) async throws -> AuthResponseDTO {
        try await client.requestFormEncoded(
            url: url("auth/login"),
            formFields: ["username": username, "password": password],
            requiresAuth: false
        )
    }
    
    func register(request: RegisterRequestDTO) async throws -> AuthResponseDTO {
        try await client.request(
            url: url("auth/register"),
            method: "POST",
            body: request,
            requiresAuth: false
        )
    }
    
    // MARK: - Users
    
    func getCurrentUser() async throws -> UserDTO {
        try await client.request(url: url("users/me"))
    }
    
    func updateCurrentUser(request: UserUpdateRequestDTO) async throws -> UserDTO {
        try await client.request(url: url("users/me"), method: "PUT", body: request)
    }
    
    func getAllUsers() async throws -> [UserDTO] {
        try await client.request(url: url("users/"))
    }
    
    func getUser(id: Int) async throws -> UserDTO {
        try await client.request(url: url("users/\(id)"))
    }
    
    func deleteUser(id: Int) async throws {
        try await client.requestVoid(url: url("users/\(id)"), method: "DELETE")
    }
    
    func adminCreateUser(request: RegisterRequestDTO) async throws -> UserDTO {
        try await client.request(url: url("users/create"), method: "POST", body: request)
    }
    
    // MARK: - Incidents
    
    func getIncidents(
        page: Int = 1,
        size: Int = 20,
        status: String? = nil,
        hazardType: String? = nil,
        urgency: String? = nil,
        search: String? = nil
    ) async throws -> IncidentListResponseDTO {
        let queryItems = [
            URLQueryItem(name: "page", value: "\(page)"),
            URLQueryItem(name: "size", value: "\(size)"),
            URLQueryItem(name: "status", value: status),
            URLQueryItem(name: "hazard_type", value: hazardType),
            URLQueryItem(name: "urgency", value: urgency),
            URLQueryItem(name: "search", value: search)
        ]
        return try await client.request(url: url("incidents/", queryItems: queryItems))
    }
    
    func getIncident(id: Int) async throws -> IncidentDTO {
        try await client.request(url: url("incidents/\(id)"))
    }
    
    func createIncident(request: CreateIncidentRequestDTO) async throws -> IncidentDTO {
        try await client.request(url: url("incidents/"), method: "POST", body: request)
    }
    
    func checkMeshMessages(request: MeshCheckRequestDTO) async throws -> MeshCheckResponseDTO {
        try await client.request(url: url("incidents/mesh/check"), method: "POST", body: request)
    }
    
    func verifyIncident(id: Int) async throws -> MessageResponseDTO {
        try await client.request(url: url("incidents/\(id)/verify"), method: "PUT")
    }
    
    func deployResponse(id: Int) async throws -> MessageResponseDTO {
        try await client.request(url: url("incidents/\(id)/deploy"), method: "PUT")
    }
    
    func resolveIncident(id: Int) async throws -> MessageResponseDTO {
        try await client.request(url: url("incidents/\(id)/resolve"), method: "PUT")
    }
    
    func assignIncident(id: Int, userId: Int) async throws -> MessageResponseDTO {
        try await client.request(
            url: url("incidents/\(id)/assign"),
            method: "PUT",
            body: ["assigned_to_id": userId]
        )
    }
    
    func getMyAssignedIncidents(page: Int = 1, size: Int = 20) async throws -> IncidentListResponseDTO {
        let queryItems = [
            URLQueryItem(name: "page", value: "\(page)"),
            URLQueryItem(name: "size", value: "\(size)")
        ]
        return try await client.request(url: url("incidents/assigned/me", queryItems: queryItems))
    }
    
    // MARK: - Analytics
    
    func getDashboardAnalytics() async throws -> DashboardAnalyticsDTO {
        try await client.request(url: url("analytics/dashboard"))
    }
    
    func getIncidentsTimeline(days: Int = 30) async throws -> IncidentsTimelineDTO {
        let queryItems = [URLQueryItem(name: "days", value: "\(days)")]
        return try await client.request(url: url("analytics/incidents/timeline", queryItems: queryItems))
    }
    
    func getIncidentsDistribution() async throws -> IncidentsDistributionDTO {
        try await client.request(url: url("analytics/incidents/distribution"))
    }
    
    func getGeographicAnalytics() async throws -> GeographicAnalyticsDTO {
        try await client.request(url: url("analytics/geographic"))
    }
    
    // MARK: - File Upload
    
    func uploadIncidentPhoto(imageData: Data, fileName: String) async throws -> UploadResponseDTO {
        try await client.uploadMultipart(
            url: url("upload/incident-photo"),
            fileData: imageData,
            fileName: fileName,
            mimeType: "image/jpeg"
        )
    }
}
